phpBB/phpbb/db/tools/doctrine.php

```diff
@@ -0,0 +1,394 @@
+<?php
+/**
+ *
+ * This file is part of the phpBB Forum Software package.
+ *
+ * @copyright (c) phpBB Limited <https://www.phpbb.com>
+ * @license GNU General Public License, version 2 (GPL-2.0)
+ *
+ * For full copyright and license information, please see
+ * the docs/CREDITS.txt file.
+ *
+ */
+
+namespace phpbb\db\tools;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Exception;
+use Doctrine\DBAL\Schema\AbstractAsset;
+use Doctrine\DBAL\Schema\Comparator;
+use Doctrine\DBAL\Schema\Index;
+use Doctrine\DBAL\Schema\Table;
+use phpbb\db\doctrine\table_helper;
+
+/**
+ * BC layer for database tools.
+ *
+ * In general it is recommended to use Doctrine directly instead of this class as this
+ * implementation is only a BC layer.
+ *
+ * In the 3.3.x version branch this class could return SQL statements instead of
+ * performing changes. This functionality has been removed.
+ */
+class doctrine implements tools_interface
+{
+	/**
+	 * @var Comparator
+	 */
+	private $comparator;
+
+	/**
+	 * @var \Doctrine\DBAL\Schema\AbstractSchemaManager
+	 */
+	private $schema_manager;
+
+	/**
+	 * Database tools constructors.
+	 *
+	 * @param Connection $connection
+	 *
+	 * @throws Exception If the schema manager cannot be created.
+	 */
+	public function __construct(Connection $connection)
+	{
+		$this->comparator = new Comparator();
+		$this->schema_manager = $connection->createSchemaManager();
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function perform_schema_changes(array $schema_changes): void
+	{
+		// @todo
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function sql_list_tables(): array
+	{
+		try
+		{
+			return array_map('strtolower', $this->schema_manager->listTableNames());
+		}
+		catch (Exception $e)
+		{
+			return [];
+		}
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function sql_table_exists(string $table_name): bool
```

Comment:
While you can make the return type more strict, that does not work for the parameter type:
```suggestion
	public function sql_table_exists($table_name): bool
```
Possible solution: Modify the interface (that's going to cause some potential BC break).

Reply:
It only breaks BC if you passed it anything but a string. And you need to pass it a string, so it should be fine.